src/socketio.quiver.js

```diff
@@ -263,51 +286,116 @@ QuiverSocialProvider.prototype.login = function(loginOpts, continuation) {
+        // failing or unreachable.  Remove it from the active list.
+        // It will then be replaced by a friend's server.
+        results.failed.forEach(this.removeServer_, this);
+      }
+      // Step 2: Connect to friends.  If I don't have MAX_CONNECTIONS long-term
+      // servers of my own, I will adopt the first new server(s) to which I
+      // connect during this process as long-term servers.
+      return connectToFriends();
+    }.bind(this)).then(function() {
+      var deficit = QuiverSocialProvider.MAX_CONNECTIONS_ - this.countOwnerConnections_();
+      // Step 3: If, after the above completes, I still have too few servers,
+      // I will retry all the default servers, and add them to the long-term
+      // set until I have MAX_CONNECTIONS servers or run out of default servers.
+      if (deficit > 0) {
+        /** @type {!Object<string, QuiverSocialProvider.server_>} */
+        var unusedDefaultServers = {};
+        QuiverSocialProvider.DEFAULT_SERVERS_.forEach(function(server) {
+          var key = QuiverSocialProvider.serverKey_(server);
+          if (!(key in this.connections_)) {
+            unusedDefaultServers[key] = server;
+          }
+        }, this);
+        return this.connectLoop_(unusedDefaultServers,
+            this.connectAsOwner_.bind(this), deficit);
+      }
+    }.bind(this)).then(function() {
+      if (this.finishLogin_) {
+        console.warn('All server connection attempts failed!');
+        // Arguably, we should report failure, not success at this point.
+        // However, the only way for the user to get back to a working state is
+        // to accept an invitation that includes a new (working) server, and
+        // they can't do that unless login has succeeded.
+        // TODO: Report failure once Quiver and application code allow us to
+        // process invitations without being logged in.
+        this.finishLogin_();
+      }
+    }.bind(this));
   }.bind(this));
 };
 
 /**
+ * @private @typedef {{
+ *   succeeded: !Array<!QuiverSocialProvider.server_>,
+ *   failed: !Array<!QuiverSocialProvider.server_>
+ * }}
+ */
+QuiverSocialProvider.connectLoopResults_ = undefined;
+
+/**
  * Async for-loop.  Try to connect to each server, and stop once we hit the
  * limit or run out of servers.
  * @param {!Object.<string, !QuiverSocialProvider.server_>} servers
  * @param {function(!QuiverSocialProvider.server_, Function)} connector
- * @return {!Promise}
+ * @param {number=} opt_limit Optional limit, defaults to MAX_CONNECTIONS
+ * @return {!Promise<!QuiverSocialProvider.connectLoopResults_>} Always fulfills.
+ *     
  * @private
  */
-QuiverSocialProvider.prototype.connectLoop_ = function(servers, connector) {
-  /** @type {Function} */ var fulfill;
-  /** @type {Function} */ var reject;
+QuiverSocialProvider.prototype.connectLoop_ = function(servers, connector,
+    opt_limit) {
+  var limit = opt_limit || QuiverSocialProvider.MAX_CONNECTIONS_;
```

Comment:
I think we do actually want a limit, because otherwise friends can force us to establish an unlimited number of connections.  However, I suppose that limit might not be 2.  Would you rather we split `MAX_CONNECTIONS_` into separate client and server limits?